### DISCUSSION TITLE
AsyncModbusTcpClient stops after ~30 seconds of inactivity

### DISCUSSION THREAD

**Discussion by RoxanaDaniela** [pymodbus-dev/pymodbus#1187] [Q&A] [answered]
Hello, 

I'm using python3-pymodbus/now 3.0.0-5 all from packages.debian.org with AsyncModbusTcpClient connected with an IP from modbus device and port=502. 

I already have a asyncio running loop and what I do is  `client = AsyncModbusTcpClient("x.x.x.x", 502) `  and ` await client.connect()` then I successfully read some registers with `await client.read_input_registers` but in a few seconds (30s of inactivity- without writing or reading registers), the client disconnects  `['ConnectionException', 'Modbus Error: [Connection] Not connected[AsyncModbusTcpClient x.x.x.x:502]']` and client.connected returns fail.

Is there any parameter to keep alive the connection or another approach?

Thank you

**Comment by janiversen** [comment]
First of all let me warn you, the version you are using have quite a number of bugs, 3.0.2 is the stable release.

Did you read the docs  ? https://pymodbus.readthedocs.io/en/v3.0.2/source/library/pymodbus.client.html#client-setup

We do not have a keep-alive option, but it might be activated pr default at socket level, I do not remember. Personally I think your problem is something else.

Pull requests are welcome.

**Reply by RoxanaDaniela** [reply]
Now I'm running with 3.0.2 and I have the same behaviour. Somehow the connection with the client closes automatically after a few seconds? It's that possible? Now I run a very simple code from the examples shown in the documentation. With 10 seconds I don't have this problem, I see the last print with End of Program.

```
from pymodbus.client import AsyncModbusTcpClient
def setup_async_client():
    client = AsyncModbusTcpClient("x.x.x.x", 502)
    return client
```
```
async def run_async_client(client):
    print("Connecting..")
    await client.connect()
    print("Client connected: {}".format(client.connected))

    reg = await client.read_input_registers(128, 1)
    print("Input Register is: {}".format(reg.registers))

    print("Waiting 30s..")
    await asyncio.sleep(30)
    print("End of Waiting")

    print("Client connected: {}".format(client.connected))
    reg = await client.read_input_registers(128, 1)
    print("Input Register is{}".format(reg.registers))
    print("Client protocol{}".format(client.protocol))
    print("End of Program")    
```
```
if __name__ == "__main__":
    testclient = setup_async_client()
    asyncio.run(run_async_client(testclient))
```

OUTPUT:
Connecting..
Client connected: True
Input Register is: [1]
Waiting 20s..
End of Waiting
Client connected: False
....
raise ConnectionException(f"Not connected[{str(self)}]")
pymodbus.exceptions.ConnectionException: Modbus Error: [Connection] Not connected[AsyncModbusTcpClient x.x.x.x:502]

**Comment by janiversen** [comment]
Please add debug as per documentation (README), otherwise I do not have a clue what happens.

Apart from the missing debug log, your example is quite nice and makes it easy to test.

**Comment by janiversen** [comment, marked as answer]
I just ran your example:
```
(pymodbus) examples % python3 connect.py
Connecting..
Client connected: True
Input Register is: [17]
Waiting 30s..
End of Waiting
Client connected: True
Input Register is[17]
Client protocolModbusClientProtocol 127.0.0.1:502
End of Program
```

Which makes me wonder if the problem is that your device disconnects. Most devices have a timeout builtin to break an unused connection. From my experience the timeout is normally around 20-30 seconds.

Add logging and you will most likely see a message that the device closed the connection.

The keep-alive option of tcp/ip will NOT help here, since it is on application level, the device needs a modbus request within its builtin timeout.

**Reply by RoxanaDaniela** [reply]
Thank you a lot, I will add logging to see what happens. 
With version 2.5.3 and sync modbus client `from pyModbusTCP.client import ModbusClient` everything works fine with same device, that's why I didn't suspect that it could be a problem with device timeout. Maybe it's different with async

**Comment by janiversen** [comment]
timeout did not really work in 2.5.3 and the timing in sync was quite different.
